### ISSUE TITLE
FBX nodes get imported with a hierarchy and wrong transforms in 5.0

### ISSUE THREAD

**Issue by ousnius** [ousnius/BodySlide-and-Outfit-Studio#269]
@sts1skj I think this is caused by the changes you did that take the nodes from the reference skeleton. I've attached the following files:
- Lenore.fbx (the FBX to import)
- Lenore 4.9.nif (the correct NIF produced by OS 4.9)
- Lenore 5.0.nif (the NIF with wrong nodes and bone transforms produced by OS 5.0)

Import the FBX into OS and save it to a NIF right away.
Do you want to take a look?

[Lenore.zip](https://github.com/ousnius/BodySlide-and-Outfit-Studio/files/4395367/Lenore.zip)

**Comment by sts1skj** [comment]
@ousnius Okay.  I'll get right on it.

**Comment by sts1skj** [comment]
@ousnius See pull request 270.  With this change, Lenore.fbx exports to a nif that looks right in NifSkope and has good skin-to-bone transforms, unlike before.

Is this everything you wanted fixed?  Your title mentions "with a hierarchy".  Recall that my transform-cleanup changes, among other things, make Outfit Studio set the appropriate node parent-child relationships when adding new bones to a nif, which is different from how Outfit Studio worked before, where it always added new bones as children of the root node.  I got the impression that this is the correct behavior when adding new bones.  I remember you specifically saying that existing parent-child relationships in the nif should be preserved.  As far as I know, having a proper hierarchy is necessary for custom bones to work properly with physics mods.  (Though, the current FBX importer does not preserve parent-child relationships for custom bones.)

**Comment by ousnius** [comment]
@sts1skj New bones that exist in the reference skeleton need to be added to the root only (no hierarchy) unless there is already a node hierarchy in the loaded NIF/project, then they need to be added to the correct parent. This is the behavior of OS in 4.9 and it needs to be that way in 5.0 (if it's not).

Custom bones not present in the reference skeleton that are added by the user need a full hierarchy so that physics mods like HDT PE and HDT SMP know where to parent the node to.

**Comment by ousnius** [comment]
@sts1skj The transforms of the FBX import are correct now, but the node hierarchy should be flat (open my comparison NIFs in NifSkope). Sorry, I merged your PR already.

**Comment by sts1skj** [comment]
I'm pretty sure OS 4.9 would _always_ add new bones as children of the root node, ignoring any existing hierarchy.  (So new custom bones would not work for physics in OS 4.9 unless you fixed them in NifSkope.)

What should the test be for whether to add a nif bone node to the root or to the bone's parent node?  It seems like a tricky question to me.  If custom bones need a full hierarchy, then should ancestors of a custom bone be reparented if they're not in a hierarchy?  Or maybe all bones should be forced to follow a hierarchy if any custom bones are present?  If no custom bones are present, should any existing hierarchy be flattened?  What if some standard bones are in a hierarchy and others are not?  What if the existing hierarchy doesn't match the structure of the reference skeleton?  (This last question is important, since each node has a bone-cs-to-parent-cs transform that is only filled in by OS for nodes whose parent is the root and for new nodes.)  What if the existing hierarchy for custom bones doesn't match the custom-bone skeleton in AnimSkeleton?

You wrote that new standard-bone nodes should have the root node as the parent "unless there is already a node hierarchy".  How do you test if there is already a node hierarchy?  What if only custom bones are already in a node hierarchy?  What if some standard bones are in a hierarchy and others aren't?

I don't know how to answer these questions well at this time.  I'll think about it.

**Comment by ousnius** [comment]
@sts1skj Yes, OS 4.9 always added custom bones to the root, which meant you couldn't use it for HDT physics until editing in NifSkope. The FBX import aren't custom nodes, though, they're known in the reference skeleton so they should be flat under the root.

Adding a custom bone, it should get added with a hierarchy. Other existing nodes do not need to be put into the hierarchy, they can stay flat. The game will still load them correctly, the hierarchy is only needed for the custom bones for the game/physics plugins to know where to parent it to.

In theory you don't need the full hierarchy either, it's enough to have a custom bone hook directly to its parent if it exists already. For example:

```
Scene Root
    -> Thigh
        -> SkirtPhysics1 (some random HDT physics bone)
    -> Calf
    -> Pelvis
```

But having it start at NPC Root and go down the full chain won't hurt either.

Take following existing NIF:
```
Scene Root
    -> Thigh
    -> Calf
    -> Pelvis
```

Now we want to add the custom bone "SkirtPhysics1" to it and we say "Thigh" is its parent. As the parent is already present somewhere in the NIF, the new node simply needs to get hooked onto it like above:

```
Scene Root
    -> Thigh
        -> SkirtPhysics1
    -> Calf
    -> Pelvis
```

Now take following existing NIF instead:
```
Scene Root
    -> Calf
    -> Pelvis
```

Adding the same custom bone, but this time we don't have its parent "Thigh" in the NIF already. Now it's enough to add "Thigh" to the root and hook "SkirtPhysics1" to it:

```
Scene Root
    -> Thigh
        -> SkirtPhysics1
    -> Calf
    -> Pelvis
```

Another example of an existing NIF that was exported in a different style:
```
Scene Root
    -> Root
        -> Pelvis
            -> Thigh
                -> Calf
                -> SkirtPhysics1
                -> SkirtPhysics2
                -> SkirtPhysics3
```

We want to add the custom bone "SkirtPhysics4" to it and choose "Thigh" as the parent. "Thigh" is already present, this time inside the hierarchy, not under the root. So we find "Thigh" and add "SkirtPhysics4" to it, which is also where the others are located.

```
Scene Root
    -> Root
        -> Pelvis
            -> Thigh
                -> Calf
                -> SkirtPhysics1
                -> SkirtPhysics2
                -> SkirtPhysics3
                -> SkirtPhysics4
```

**Comment by sts1skj** [comment]
@ousnius (By the way, OutfitProject::ImportFBX _does_ have code for importing custom bones, but it doesn't work.)

You seem to be telling me the following rule for adding a custom bone node to a NIF file: attach it to its parent bone's node if that exists; if not, recursively add ancestors until you get to the root or to an ancestor bone node that does exist.  Is that correct?

You also seem to be saying that standard (exists in skeleton) bones with nodes in the NIF should not have those nodes reparented.

But what about when adding a new standard-bone node to a NIF file?  Should some test be performed to see if the NIF's existing structure is flat or not?  What if the only non-flat parts are custom bones?  Or should new standard-bone nodes _always_ be added to the root, ignoring existing hierarchy?

What if a custom bone's parent in AnimSkeleton disagrees with its parent in the NIF file?  Should it be reparented?  Right now, the OutfitProject functions AddCustomBoneRef and ModifyCustomBone do not change the NifFile, so they can easily be out of sync.  If someone fixes ImportFBX so it preserves the custom bone's parent attachment, that could be out of sync with the NifFile too.

Should the node-to-parent transform present in every NIF node be set correctly for _all_ bone nodes?  Currently, OS only sets it (in AnimInfo::WriteToNif) for new bones and nodes whose parent is the root (so standard bones in a hierarchy are the only ones not updated).

**Comment by ousnius** [comment]
> You seem to be telling me the following rule for adding a custom bone node to a NIF file: attach it to its parent bone's node if that exists; if not, recursively add ancestors until you get to the root or to an ancestor bone node that does exist. Is that correct?

Correct

> You also seem to be saying that standard (exists in skeleton) bones with nodes in the NIF should not have those nodes reparented.

Correct

> But what about when adding a new standard-bone node to a NIF file? Should some test be performed to see if the NIF's existing structure is flat or not? What if the only non-flat parts are custom bones? Or should new standard-bone nodes always be added to the root, ignoring existing hierarchy?

New standard bones are always added to the root. They're part of the reference skeleton and game's skeleton, so the hierarchy is not needed, even when one is present in the NIF already.

> What if a custom bone's parent in AnimSkeleton disagrees with its parent in the NIF file?

I don't understand, the definition of a custom bone is that it's not in AnimSkeleton (reference skeleton).

> Should the node-to-parent transform present in every NIF node be set correctly for all bone nodes? Currently, OS only sets it (in AnimInfo::WriteToNif) for new bones and nodes whose parent is the root (so standard bones in a hierarchy are the only ones not updated).

I can't think of any reason that would stop us from updating all nodes, provided what we update them with is correct (reference skeleton and hierarchy in mind).

**Comment by sts1skj** [comment]
(In case it's not obvious from my flailing around, I still don't really understand Outfit Studio's project data model.)

>     What if a custom bone's parent in AnimSkeleton disagrees with its parent in the NIF file?
> 
> I don't understand, the definition of a custom bone is that it's not in AnimSkeleton (reference skeleton).

I think I've figured out the answer to this question on my own: the custom bone's node should be reparented in the NIF file.  The rest of this message tries to explain why; feel free to skip it if you don't care.

Custom bones are stored in AnimSkeleton::customBones, while standard bones (reference skeleton bones) are stored in AnimSkeleton::allBones.  Access functions like AnimSkeleton::GetBonePtr search both lists by default.  Functions like AnimInfo::LoadFromNif, OutfitProject::ImportFBX, and OutfitProject::AddCustomBoneRef store custom bones there.  This isn't new behavior; it was like this before I touched anything.  Sometimes custom bone data is also stored in various lib/NIF data structures, but not always.

My question was focusing on the function AnimInfo::WriteToNif, which is the main function for transferring bone data (among other things) from the various Anim.h data structures into the various lib/NIF data structures.  Most of the changes we're discussing in this thread will be made to that function.

So the question I was asking was: what if, in AnimInfo::WriteToNif, the existing custom bone's parent in NifFile disagrees with its parent in AnimSkeleton/AnimBone?  Should the NifFile's bone node be reparented?

It seems clear to me now that the answer is "yes, reparent it".  Various functions update the custom bone's AnimBone data structure (in response to user input, and possibly also when loading files) but don't update the corresponding data in NifFile.  (Maybe they should.)  So WriteToNif should make the custom bone's parent in NifFile match what it is in AnimBone.

**Comment by ousnius** [comment]
Yeah, especially since you can reparent them in the UI now.

**Comment by sts1skj** [comment]
So here's my proposed set of changes to implement what I think you want:

1.  Add code to AnimInfo::WriteToNif to handle new standard bone nodes separately from new custom bone nodes.  New custom bone nodes will just be directly added with NifFile::AddNode (with parent set to nullptr).

2.  Add a check to AnimInfo::WriteToNif to make sure every custom bone's parent, as returned by NifFile::GetParentNode, matches AnimBone::parent.  If not, use some unknown NifFile function to change the parent node.

3.  Change the code in AnimInfo::WriteToNif that calls NifFile::SetNodeTransformToParent so it's always done (instead of only for children of the root), and calculate a good transform to pass instead of using AnimBone::xformToGlobal.

Shall I start implementing this?

**Comment by ousnius** [comment]
Yeah, you can.

> New custom bone nodes will just be directly added with NifFile::AddNode (with parent set to nullptr).

I'm assuming you mean new standard nodes? Because custom nodes need to be under the parent that was chosen in the dialog.

**Comment by sts1skj** [comment]
Yes, you're right: I should have written "new standard bone nodes".

**Comment by ousnius** [comment]
This is fixed with 1f5fffb.
